docs/proposals/client-package-structure.md

```diff
@@ -0,0 +1,153 @@
+current version inspecific types.
+
+We will provide this structure for each version of each group. It is infeasible
+to do this manually, so we will generate this. The generator will accept both
+swagger and the ordinary go types.
+
+Typed clients will be constructabale given a MetaClient.
+
+Action items:
+* Move current typed clients into new directory structure (described below)
+* Finish client generation logic. (@lavalamp)
+
+#### High-level, typed: Discovery
+
+A `DiscoveryClient` is necessary to discover the api groups, versions, and
+resources a server supports. It's constructable given a RESTClient. It is
+consumed by both the MetaClient and users who want to iterate over groups,
+versions, or resources. (Example: namespace controller.)
+
+Action items:
+* Refactor existing functions to present a sane interface, as close to that
+  offered by the other typed clients as possible. (@caeserxuchao)
+
+### High-level: Client Sets
+
+Because there will be multiple groups with multiple versions, we will provide an
+aggregation layer that combines multiple typed clients in a single object. We do
+this to constrain the testing matrix.
+
+A client set is constructable given a MetaClient.
+
+An example:
+
+```go
+import (
+  api_v1 "k8s.io/kubernetes/pkg/client/typed/generated/v1"
+  ext_v1beta1 "k8s.io/kubernetes/pkg/client/typed/generated/extensions/v1beta1"
+  net_v1beta1 "k8s.io/kubernetes/pkg/client/typed/generated/net/v1beta1"
+  net_v1beta1 "k8s.io/kubernetes/pkg/client/typed/generic"
+)
+
+type Client interface {
+  API() api_v1.Client
+  Extensions() ext_v1beta1.Client
+  Net() net_v1beta1.Client
+
+  GroupVersion(group, version string) generic.Client
+}
+```
+
+Note that a particular version is chosen for each group. It is a general rule
+for our API structure that no client need care about more than one version of
+each group at a time.
+
+This is the primary deliverable that people would consume. It is also generated.
+
+Action items:
+* This needs to be built. It will replace the ClientInterface that everyone
+  passes around right now.
+
+### High-level: Generic
+
+The generic client lets users consume apis which are not compiled into their
+binary. It will provide the same interface as the typed client, but will take
+and return `runtime.Object`s instead of typed objects. There is only one generic
+client, so it's not necessary to generate it, although optionally we may do so
+depending on whether the typed client generator makes it easy.
+
+Action items:
+* A rudimentary version of this exists in kubectl's builder. It needs to be
+  moved to a more general place.
+
+
+## Package Structure
+
+```
+pkg/client/
+----------/transport/     # transport & associated config
+----------/metaclient/
+----------/typed/
+----------------/discovery/
+----------------/generated/
+--------------------------/<group>/<version>/<resource>.go
+--------------------------/<setname>/
+----------------/generic/
+----------/clientset/
+```
+
```

Comment:
Other steps:

* use versioned param structs (list options, delete options) everywhere
* remove Kube API group specific behavior from request.go
* formalize SetKubeDefaults with didcover, remove any overlap.  Discovery should probably only occur when apiversion is unspecified on config